src/marks/raster.js

```diff
@@ -0,0 +1,437 @@
+      FO = new Array(w * h);
+      for (let yi = 0, i = 0; yi < h; ++yi) {
+        for (let xi = 0; xi < w; ++xi, ++i) {
+          FO[i] = fillOpacity(x.invert(x1 + xi * kx), y.invert(y1 + yi * ky));
+        }
+      }
+    }
+    return {
+      data: F ?? FO,
+      facets,
+      channels: {
+        ...(F && {fill: {value: F, scale: "color"}}),
+        ...(FO && {fillOpacity: {value: FO, scale: "opacity"}})
+      }
+    };
+  });
+}
+
+function maybeRasterize(rasterize) {
+  if (typeof rasterize === "function") return rasterize;
+  if (rasterize == null) return rasterizeNone;
+  switch (`${rasterize}`.toLowerCase()) {
+    case "none":
+      return rasterizeNone;
+    case "dense":
+      return rasterizeDense;
+    case "nearest":
+      return rasterizeNearest;
+    case "barycentric":
+      return rasterizeBarycentric;
+    case "random-walk":
+      return rasterizeRandomWalk;
+  }
+  throw new Error(`invalid rasterize: ${rasterize}`);
+}
+
+// Applies a simple forward mapping of samples, binning them into pixels without
+// any blending or interpolation.
+function rasterizeNone(canvas, index, {color}, {fill: F, fillOpacity: FO}, {x: X, y: Y}) {
+  const {width, height} = canvas;
+  const context2d = canvas.getContext("2d");
+  const image = context2d.createImageData(width, height);
+  const imageData = image.data;
+  let {r, g, b} = rgb(this.fill) ?? {r: 0, g: 0, b: 0};
+  let a = 255;
+  for (const i of index) {
+    if (X[i] < 0 || X[i] >= width || Y[i] < 0 || Y[i] >= height) continue;
+    const j = (Math.floor(Y[i]) * width + Math.floor(X[i])) << 2;
+    if (F) ({r, g, b} = rgb(color(F[i])));
+    if (FO) a = FO[i] * 255;
+    imageData[j + 0] = r;
+    imageData[j + 1] = g;
+    imageData[j + 2] = b;
+    imageData[j + 3] = a;
+  }
+  context2d.putImageData(image, 0, 0);
+}
+
+// Assumes that the fill and/or fillOpacity channels are a dense grid of values
+// that correspond to the size of the given canvas, in row-major order.
+function rasterizeDense(canvas, index, {color}, {fill: F, fillOpacity: FO}) {
+  const {width, height} = canvas;
+  const context2d = canvas.getContext("2d");
+  const image = context2d.createImageData(width, height);
+  const imageData = image.data;
+  let {r, g, b} = rgb(this.fill) ?? {r: 0, g: 0, b: 0};
+  let a = 255;
+  for (let i = 0, n = width * height; i < n; ++i) {
+    const j = i << 2;
+    if (F) {
+      const fi = color(F[i]);
+      if (fi == null) {
+        imageData[j + 3] = 0;
+        continue;
+      }
+      ({r, g, b} = rgb(fi));
+    }
+    if (FO) a = FO[i] * 255;
+    imageData[j + 0] = r;
+    imageData[j + 1] = g;
+    imageData[j + 2] = b;
+    imageData[j + 3] = a;
+  }
+  context2d.putImageData(image, 0, 0);
+}
+
+function rasterizeNearest(canvas, index, {color}, {fill: F, fillOpacity: FO}, {x: X, y: Y}) {
+  const {width, height} = canvas;
+  const context = canvas.getContext("2d");
+  const voronoi = Delaunay.from(
+    index,
+    (i) => X[i],
+    (i) => Y[i]
```

Comment:
```suggestion
    (i) => X[i] + 0.5,
    (i) => Y[i] + 0.5
```

Reply:
I think it’s correct as-is. You can see a noticeable drift with large pixelSize if you make this change.